### PR TITLE
Some modification on CRF

### PR DESCRIPTION
**Pull request by linxihui**
1. Separated `loss` and accuracy metrics from class instance, so that `keras.models.load_model` works with CRF.
2. Fixed problem with `bias=False`.

### CODE REVIEW

File: keras_contrib/layers/crf.py [hunk 1]
```diff
@@ -282,63 +289,19 @@ def get_config(self):
...
-                mask = self._inbound_nodes[0].input_masks[0]
-                nloglik = self.get_negative_log_likelihood(y_true, X, mask)
-                return nloglik
-            return loss
-        else:
-            if self.sparse_target:
-                return sparse_categorical_crossentropy
-            else:
-                return categorical_crossentropy
+        raise AttributeError('"CRF.loss_function" is depreciated. Please use "losses.crf_loss".')
```
Comment (titu1994) [inline keras_contrib/layers/crf.py]: Instead of raising an Error, is it possible to determine the appropriate loss and supply it here? If not, then this is fine.

File: keras_contrib/layers/crf.py [hunk 2]
```diff
@@ -282,63 +289,19 @@ def get_config(self):
...
-            y_true = K.cast(y_true[:, :, 0], K.dtype(y_pred))
-        else:
-            y_true = K.argmax(y_true, -1)
-        judge = K.cast(K.equal(y_pred, y_true), K.floatx())
-        if mask is None:
-            return K.mean(judge)
-        else:
-            mask = K.cast(mask, K.floatx())
-            return K.sum(judge * mask) / K.sum(mask)
+        raise AttributeError('"CRF.accuracy" is depreciated. Please use "metrics.crf_accuracy".')
```
Comment (titu1994) [inline keras_contrib/layers/crf.py]: Same as above

File: keras_contrib/layers/crf.py [hunk 3]
```diff
@@ -282,63 +289,19 @@ def get_config(self):
...
     @property
     def viterbi_acc(self):
-        def acc(y_true, y_pred):
-            X = self._inbound_nodes[0].input_tensors[0]
-            mask = self._inbound_nodes[0].input_masks[0]
-            y_pred = self.viterbi_decoding(X, mask)
-            return self._get_accuracy(y_true, y_pred, mask, self.sparse_target)
-        acc.func_name = 'viterbi_acc'
-        return acc
+        raise AttributeError('"CRF.viterbi_acc" is depreciated. Please use "metrics.crf_viterbi_accuracy".')
```
Comment (titu1994) [inline keras_contrib/layers/crf.py]: Similar to above. Can this be inferred and the correct value be returned here itself instead of raising an Error?

File: keras_contrib/layers/crf.py [hunk 4]
```diff
@@ -282,63 +289,19 @@ def get_config(self):
...
     @property
     def marginal_acc(self):
-        def acc(y_true, y_pred):
-            X = self._inbound_nodes[0].input_tensors[0]
-            mask = self._inbound_nodes[0].input_masks[0]
-            y_pred = self.get_marginal_prob(X, mask)
-            return self._get_accuracy(y_true, y_pred, mask, self.sparse_target)
-        acc.func_name = 'marginal_acc'
-        return acc
+        raise AttributeError('"CRF.marginal_acc" is depreciated. Please use "metrics.crf_marginal_accuracy".')
```
Comment (titu1994) [inline keras_contrib/layers/crf.py]: Same as above

File: tests/keras_contrib/layers/test_crf.py [hunk 5]
```diff
@@ -4,8 +4,12 @@
 
 from keras.utils.test_utils import keras_test
 from keras.layers import Embedding
+from keras.models import Sequential, model_from_json, load_model
+from keras.models import load_model
+from keras_contrib.losses import crf_loss
+from keras_contrib.metrics import crf_accuracy, crf_marginal_accuracy, crf_viterbi_accuracy
 from keras_contrib.layers import CRF
-from keras.models import Sequential, model_from_json
+from keras_contrib import *
```
Comment (titu1994) [inline tests/keras_contrib/layers/test_crf.py]: Please do not use wildcard imports. Import only what is needed.

**Comment by lzfelix** [comment]
Will this be merged on the future?

**Comment by titu1994** [comment]
When the issues raised above are resolved.

**Comment by lzfelix** [comment]
My apologies, I thought that they had already been fixed.

**Comment by titu1994** [comment]
No issues. I feel like this should be merged as soon as possible, since it is a cleaner way of doing things.

File: tests/keras_contrib/layers/test_crf.py (hunk 5 above)
Comment (linxihui) [reply tests/keras_contrib/layers/test_crf.py]: This is suggested by the `keras-contrib`  github home page, to handle `load_model` issue. But it does not seem to work.

Comment (linxihui) [reply tests/keras_contrib/layers/test_crf.py]: I cannot manage to successfully using `keras.models.load_model` to load model from `keras-contrib`.  If you have a working example,  would you mind sharing with me?

Comment (titu1994) [reply tests/keras_contrib/layers/test_crf.py]: Hmm, I haven't used it much myself. Simply passing the custom objects should be fine, but I've never tried it.

Comment (lzfelix) [reply tests/keras_contrib/layers/test_crf.py]: @linxihui There is a workaround for the CRF layer here on #125.

Comment (titu1994) [reply tests/keras_contrib/layers/test_crf.py]: This is the only issue remaining before I merge the PR. Is there no workaround than to do a wildcard import?

Comment (kamei86i) [reply tests/keras_contrib/layers/test_crf.py]: The solution/workaround I posted here (#129 ) seems to work. I hope it could be useful to @linxihui

**Comment by lzfelix** [comment]
This PR seems stale for too long. Despite being very useful for the community, since there are multiple issues regarding the dificulties in persisting and restoring models that rely on this layer, a small modification is hindering the merge.

I took the liberty to clone this PR locally on my env, merge it with the most recent keras-contrib code (no conflicts) and solved the issue. @titu1994 and @linxihui, is it fine if I take this PR forward?

Thanks,
Luiz

**Comment by linxihui** [comment]
@lzfelix Absolutely fine for me.  Thanks for taking care of it.

**Comment by lzfelix** [comment]
Since I didn't know how to make commits in this PR, I've created a new one, #318. Please let me know if there's anything that I can do to merge them.

**Comment by gabrieldemarmiesse** [comment]
Closing in favor of #318
